File: 백준/Gold/24042. 횡단보도/횡단보도.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    static int N, M;

    static List<Node>[] edges;


    static class Node {
        int end;
        long time;

        Node(int end, long time) {
            this.end = end;
            this.time = time;
        }
    }


    public static void main(String[] args) throws IOException {

        input();

        long[] dist = new long[N + 1];
        PriorityQueue<Node> pq = new PriorityQueue<>(new Comparator<Node>() {
            @Override
            public int compare(Node o1, Node o2) {
                return (int) (o1.time - o2.time);
            }
        });

        long INF = Long.MAX_VALUE;
        Arrays.fill(dist, INF);
        dist[1] = 0;
        pq.add(new Node(1, 0));


        while (!pq.isEmpty()) {
            Node now = pq.poll();
            if(dist[now.end]<now.time) continue;

            for (Node next : edges[now.end]) {
//                long time  = (next.time - now.time)%M;
//                if(time<0) time+= M;

                long time = now.time % M;
                time = next.time - time;
                if(time<0) time+= M;

                long nextTime = now.time + time;
                if(dist[next.end]>nextTime){
                    dist[next.end] = nextTime;
                    pq.add(new Node(next.end, dist[next.end]));

                }


            }
        }

        System.out.println(dist[N]);


    }

    public static void input() throws IOException {

        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        edges = new ArrayList[N + 1];
        for (int n = 1; n <= N; n++) {
            edges[n] = new ArrayList<>();
        }
        for (int m = 1; m <= M; m++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            edges[a].add(new Node(b, m));
            edges[b].add(new Node(a, m));
        }


    }
}

File: 백준/Gold/24042. 횡단보도/횡단보도.py
import heapq
n, m = map(int, input().split()) #n 노드수 m 횡단보도 수(간선)
edges = [[] for i in range(n+1)]
for i in range(1, m+1):
    st, ed = map(int, input().split())
    edges[st].append((i, ed))
    edges[ed].append((i, st))

#시작점 pq에 넣기
#dist 시작점 0
pq = [(0, 1)] #cost / node (node는 1부터 ~ n 이므로)
inf = 700_000*2*n
dist = [inf]*(n+1)
dist[1] = 0

#pq에서 하나씩 꺼내오기
#cost값이 dist보다 크면 continue
#edge 돌기
#edge 값이 cur_cost 보다 작으면 module 더해주기( (cur_cost - next_cost)//M+1번만큼 )
#위에서 더해서 구한 값과 dist 비교후 갱신
#pq에 넣기

while pq:
    cur_cost, cur_node = heapq.heappop(pq)
    if dist[cur_node] < cur_cost:
        continue

    for e in edges[cur_node]:
        next_cost, next_node = e
        if next_cost<cur_cost:
            next_cost += ((cur_cost-next_cost)//m + 1)*m

        if dist[next_node] > next_cost:
            dist[next_node] = next_cost
            heapq.heappush(pq, (next_cost, next_node))
print(dist[n])